pkg/upgrade/upgrader/jiva_upgrade.go

```diff
@@ -630,6 +646,148 @@ func jivaUpgrade(pvName, openebsNamespace string, utaskObj *utask.UpgradeTask) (
+				RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0].MatchExpressions[0].Values = []string{node}
+			klog.Infof("creating replica deployment %s in %s namespace", replicaDeploy.Name, openebsNamespace)
+			replicaDeploy, err := deployClient.WithNamespace(openebsNamespace).Create(replicaDeploy)
+			if err != nil {
+				return err
+			}
+			err = waitUntilPodsAreRunning(openebsNamespace, replicaDeployLabel, i+1)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (j *jivaVolumeOptions) migrateTarget(openebsNamespace string) error {
+	if j.controllerObj.version == currentVersion {
+		deployObj, err := deployClient.WithNamespace(j.ns).Get(j.controllerObj.name)
+		if err != nil {
+			return err
+		}
+		var one int32 = 1
+		deployObj.Spec.Replicas = &one
+		deployObj.Namespace = openebsNamespace
+		deployObj.ResourceVersion = ""
+		// updating the controller deploymentsimage and version before
+		// creating them in openebs namespace.
+		lastIndex := strings.LastIndex(deployObj.Spec.Template.Spec.Containers[0].Image, ":")
+		deployObj.Spec.Template.Spec.Containers[0].Image = deployObj.Spec.
+			Template.Spec.Containers[0].Image[:lastIndex+1] + imageTag
+		lastIndex = strings.LastIndex(deployObj.Spec.Template.Spec.Containers[1].Image, ":")
+		deployObj.Spec.Template.Spec.Containers[1].Image = deployObj.Spec.
+			Template.Spec.Containers[1].Image[:lastIndex+1] + imageTag
+		// if target-affinity is set for the pvc them openebs namespace
+		// needs to be added as a bug fix.
+		if deployObj.Spec.Template.Spec.Affinity != nil {
+			if deployObj.Spec.Template.Spec.Affinity.PodAffinity.
+				RequiredDuringSchedulingIgnoredDuringExecution[0].LabelSelector.
+				MatchExpressions[0].Key == "openebs.io/target-affinity" {
+				deployObj.Spec.Template.Spec.Affinity.PodAffinity.
+					RequiredDuringSchedulingIgnoredDuringExecution[0].
+					Namespaces = []string{j.ns}
+			}
+		}
+
+		deployObj.Labels["openebs.io/version"] = upgradeVersion
+		klog.Infof("creating controller deployment %s in %s namespace", deployObj.Name, openebsNamespace)
+		deployObj, err = deployClient.WithNamespace(openebsNamespace).Create(deployObj)
+		if err != nil {
+			return err
+		}
+		err = waitUntilPodsAreRunning(openebsNamespace, ctrlDeployLabel, 1)
+		if err != nil {
+			return err
+		}
+		svcList, err := serviceClient.WithNamespace(j.ns).List(metav1.ListOptions{
+			LabelSelector: ctrlSVCLabel,
+		})
+		if err != nil {
+			return err
+		}
+		svcObj := &svcList.Items[0]
+		svcObj.Namespace = openebsNamespace
+		svcObj.ResourceVersion = ""
+		svcObj.Labels["openebs.io/version"] = upgradeVersion
+		klog.Infof("removing controller service %s in %s namespace", svcObj.Name, j.ns)
+		err = serviceClient.WithNamespace(j.ns).Delete(svcObj.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		klog.Infof("creating controller service %s in %s namespace", svcObj.Name, openebsNamespace)
```

Comment:
what if upgrade job dies after deleting service? we will be losing the IP..
is it possible to create service with same IP but in disabled mode?
If so, first create in disabled mode, delete old one and then enable new one?